Consolidated Change Proposal

adrianhopebailie opened the pull request:
Per discussion in the DA and on the mid-PI review this is a consolidated proposal of all changes in the issue list.

It includes fixes for:
 1. alignment with ILPv4
 2. better cryptographic binding
 3. extensions to facilitate cross-currency
 4. extensions to facilitate data collection for cross-border

ehenrka commented:
Please find some initial comments below:

**`Participant` data model:**

> rate | 0..1 | Amount | The rate of exchange that will applied by this participant. 

- How will you know what currrency rate this means as there is only a `transferCurrency` in the data model? It seems strange to have to look at the earlier Participant object in the ordered list to get the currency pair.
- I do not want to use `Amount` data type as it only allows 4 decimals (due to no currencies using more than 4). FX rates can use more than 4 decimals. A new Rate (or similar) data type should be defined. The rate data type could then also include the currency pair (see above).

> dataRequired | 0..32 | String(1..128) | List of required

- Some more descriptive text needed than just "List of required".
- For consistency reasons, this should be a `RequiredDataList `object (or similar) which then contains the the 0..32 objects of the required data (I don't know why we started using this method..).
- Shouldn't this be an enum stating which type of data that could be requested? Just having a string which could be any type of data to request seems very hard to implement in interoperable way.

> dataProvided | 0..32 | EncryptedData | The data provided by other participants for use by this participant.

- For consistency reasons, this should be a `ProvidedDataList `object (or similar) which then contains the the 0..32 objects of the provided data.

**Participant fees vs existing fees**

There are fees in the `Participant` data model now, how are these related to the existing fees? It would be great to see some end to end examples of how these fees should be handled. These fees will have a rather large implication on Section 5.1 in the API Definition document. If the Payer and Payee FSP are considered `Participant`s (otherwise they cannot request and/or provide regulatory data), should their fees then not be present in their `Participant` objects as they already have fees in the existing data model?

**`POST /quotes` data model**

> fees | 0..1 | Money | "Fees in the transaction.

- The `fees` in the `POST /quotes` service should be renamed to `payerFspFee` as there are many different fees now, or even removed and have Payer FSP as a `Participant`.

> participants | 0..16 | Participant | The participants in the transaction.

- For consistency reasons, this should be a `ParticipantList`object (or similar) which then contains the the 0..16 objects, see for example `ExtensionList`.

**`PUT /quotes` data model**

- There should be a `Transaction` object in the `PUT /quote` as the `IlpPacket` has been removed?

> transferAmount | 1 | Money | The amount of Money that the Payer FSP should transfer to the Payee FSP.

- Is "The amount of Money that the Payer FSP should transfer to the Payee FSP" really correct with the proposal? I would assume that  "The amount of Money that the Payer FSP should transfer" would be more correct if there are Participant fees that will be subtracted from the amount by intermediary `Participant`s (see also above question with Participant fees vs existing fees).

> participants | 0..16 | Participant | The participants in the transaction.

- For consistency reasons, this should be a `ParticipantList`object (or similar) which then contains the the 0..16 objects, see for example `ExtensionList`.

> echoData | 0..1 | String(1..2048) | Opaque data provided by the payee that must be echoed back unchanged in the transfer.

- For consistency reasons, the String (1..2048) should be its own data type called EchoData.

**`POST /transfers` data model**

> participants | 0..16 | Participant | The participants in the transaction.

- For consistency reasons, this should be a `ParticipantList`object (or similar) which then contains the the 0..16 objects, see for example `ExtensionList`.

> echoData | 0..1 | String(1..2048) | Opaque data provided by the payee that must be echoed back unchanged in the transfer.

- For consistency reasons, the String (1..2048) should be its own data type called EchoData.

elnyry commented on proposals/adrianhopebailie-201908.md:
```diff
@@ -0,0 +1,588 @@
+---
+Change Request ID: ???
```
Can we use the ID of the issue https://github.com/mojaloop/mojaloop-specification/issues/18 ?

adrianhopebailie commented:
@ehenrka 

> How will you know what currency rate this means as there is only a transferCurrency in the data model? It seems strange to have to look at the earlier Participant object in the ordered list to get the currency pair.

I had anticipated that you'd look at transferCurrency of the adjacent participant. Otherwise this seems like redundant information.

What do you do if the rate is expressed as 0.11111 USD/EUR but the transferCurrency of the previous participant is not USD or the transferCurrency of the current participant is not EUR?

> A new Rate (or similar) data type should be defined. The rate data type could then also include the currency pair (see above).

Good idea. I am not opposed to adding the currency pair but we should define what happens if the pair doesn't correspond to the transferCurrency of the current and previous participants. (See above)

> Some more descriptive text needed than just "List of required".

Yes, this got truncated somehow. It should be "List of required data for compliance" or similar.

> For consistency reasons, this should be a RequiredDataList object (or similar) which then contains the the 0..32 objects of the required data (I don't know why we started using this method..).

👍 

> Shouldn't this be an enum stating which type of data that could be requested? Just having a string which could be any type of data to request seems very hard to implement in interoperable way.

Yes, however we don't have a good idea yet of what these would be. I expect this will be based on some data model defined elsewhere and we'll need to define behaviour if a value is encountered that is not understood.

> For consistency reasons, this should be a ProvidedDataList object (or similar) which then contains the the 0..32 objects of the provided data.

👍 

> There are fees in the Participant data model now, how are these related to the existing fees? It would be great to see some end to end examples of how these fees should be handled. These fees will have a rather large implication on Section 5.1 in the API Definition document. If the Payer and Payee FSP are considered Participants (otherwise they cannot request and/or provide regulatory data), should their fees then not be present in their Participant objects as they already have fees in the existing data model?

These are intended to capture fees from intermediaries only. You are right that we should explicitly state that the payer FSP and payee FSP can't specify a fee or exchange rate.

> The fees in the POST /quotes service should be renamed to payerFspFee as there are many different fees now, or even removed and have Payer FSP as a Participant.

Interesting! I hadn't considered removing the fees from the existing model entirely. I agree that we need to more clearly define how the fees in the participants list and the fees in the `POST /quotes` request interact and which fees are captured where.

I'll need some more real-world experience from others on the best way to do this. Something we can discuss in London.

For consistency reasons, this should be a ParticipantListobject (or similar) which then contains the the 0..16 objects, see for example ExtensionList.

👍 

> There should be a Transaction object in the PUT /quote as the IlpPacket has been removed?

👍  good catch, I thought I had that in

> Is "The amount of Money that the Payer FSP should transfer to the Payee FSP" really correct with the proposal? I would assume that "The amount of Money that the Payer FSP should transfer" would be more correct if there are Participant fees that will be subtracted from the amount by intermediary Participants (see also above question with Participant fees vs existing fees).

👍 your wording is more correct

> For consistency reasons, this should be a ParticipantListobject (or similar) which then contains the the 0..16 objects, see for example ExtensionList.

👍 

> For consistency reasons, the String (1..2048) should be its own data type called EchoData.
POST /transfers data model

👍 

> For consistency reasons, this should be a ParticipantListobject (or similar) which then contains the the 0..16 objects, see for example ExtensionList.

👍 

> For consistency reasons, the String (1..2048) should be its own data type called EchoData.

👍 

I'll make these changes now and send an updated proposal around in preparation for the workshop.

ehenrka commented:
Thanks @adrianhopebailie.

Regarding including the actual currency pair, we could add some text describing that the pair can/should be validated against the transferCurrency of the adjacent participant. It should ideally generate an error that is returned from the next hop to be able to fix this coding error where an errorenous currency pair was added.

The following is a suggestion for a regular expression of the FX rate: `^(([1-9][0-9]{0,5}))([.][0-9]{0,8}[1-9])?$`.

This would allow the FX rate for currency pairs such as USD/VEF, for which the current rate is 18540.6793, and PHP/USD, for which the current rate is 0.01950839. An extra digit is also allowed both ways to those examples to future proof it a bit, meaning up to 6 digits in the major unit and up to 9 digits in the minor unit. It will not allow any leading or trailing zeros (same as in Amount). A single 0 is not allowed as it does not make sense to have an FX rate of 0 (this is allowed in the Amount).

adrianhopebailie commented:
> Are there any differences in this proposal compared to #22? I could not find any differences on a quick look at least.

No, commit bb459d9 was just a rebase. Have now added some commits addressing your comments. Busy working on some more to add some changes based on the London discussion.

lewisdaly commented on proposals/adrianhopebailie-201908.md:
```diff
@@ -0,0 +1,639 @@
+provided in the transfer. This is different to the Open FSP API which requires a
+quote cycle preceding a transaction.
+
+When using the API routing of a multi-hop transaction is determine during the
```
Typo: should be `"determined"`

lewisdaly commented on proposals/adrianhopebailie-201908.md:
```diff
@@ -0,0 +1,639 @@
+#### Regulatory Data
+
+When preparing a quote, each participant will indicate the data that it requires
+by putting the names data fields into the `dataRequired` list. It MUST also
```
minor semantic issue: should this be `requiredData` instead? I know there is a `RequiredData` model, so I found this field name a little confusing.

lewisdaly commented on proposals/adrianhopebailie-201908.md:
```diff
@@ -0,0 +1,639 @@
+
+#### Add `RequiredDataList`, `ProvidedDataList` and `ProvidedData`
+
+The `RequiredDataList` data model is defined as:
```
In the case of multi hops, don't we need a way to keep track of _which participant_ is requesting the required data? Or is the plan to keep adding `RequiredData` to the list as the transaction keeps accumulating hops during the Quote request "leg" of the transfer?

adrianhopebailie replied:
The list exists for each participant. I.e. Each participant adds to the list of participants which an entry for itself

lewisdaly replied:
Cool. Thanks for clarifying this for me.

ehenrka commented:
@adrianhopebailie Thank you for the updates, please find some additional comments below. Please also note that some of my original comments in https://github.com/mojaloop/mojaloop-specification/pull/22#issuecomment-527771245 are not yet updated (for example "The fees in the **POST /quotes** service should be renamed to `payerFspFee `as there are many different fees now, or even removed and have Payer FSP as a Participant.").

> **Generation of the Condition and Fulfillment**
...
This data element is a variable length UTF-8 string with a maximum length of 10 KB. This allows the payee to encode any data of their choosing in the field.

Is there a need to have 10 KB as max? In a bulk, this could become a very big request. 1000 elements * 10 KB = 10 MB just for the echo data. In the actual data element, the echoData is defined as
`String(1..2048)`, which should be enough?

> This proposal defines a new data element; EchoData that is used as the input to this algorithm and mandates that this data is passed in the PUT /quote response callback and the POST /transfer request.
...
> The API mandates (to ensure interoperability) that the EchoData received by a payer FSP in a PUT /quote response callback is sent unmodified in the corresponding POST /transfers request.

The two paragraphs above seems very similar. Suggest to combine them.

> | Output        | `PUT /quote` response callback |
> | ------------- | ------------------------------ |
> | **condition** | `PUT /quote#condition`|
> | **echoData**  | `PUT /quote#echoData`|

`PUT /quote` hould be `PUT /quotes` (missing 's'). Same typo exists in other places as well (for example in the two similar paragraphs mentioned above), just do a search and replace all instances of `/quote` with `/quotes`.

> **Update Party**

To me it seems beneficial to return the actual route as well in the Party lookup, so that the Payer FSP can decide which quotes to initiate on basis of the route information. The route information could be a simplification of the `ParticipantList `used in the quote, just containing the involved FSPs in the route which was used to find the Payee. Maybe there are FSPs in some countries that the Payer FSP would like to avoid, or the number of participants is considered to many for the Payer FSP to even bother sending a quote if there are alternative routes.

The route information from the Party would also help out with routing in the `POST /quotes`, as the "base" route should then already be known (for example FX providers could be involved in the actual quote route as the party lookup does not contain any currency information).

> **Add AccountList and Account data types**
> ...
> The AccountAddress data type is a variable length string with a maximum size of 1023 characters and consists of:
> - Alphanumeric characters, upper or lower case. (Addresses are case-sensitive so that they can contain data encoded in formats such as base64url.)
> - Underscore (_)
> - Tilde (~)
> - Hyphen (-)
> - Period (.) Addresses MUST NOT end in a period (.) character

Is the above correct with the format? If I can remember correctly, you mentioned in the CCB that the AccountAddress should be opaque?

> **Specify accountAddress in the Quote and Transfer to assist with routing**
> If a sending DFSP has an AccountAddress for the payee account this should be included in the POST /quotes request.

There is no notion of a 'sending DFSP' in the API, as sending could mean both sending of message and sending of money, which might not be the same. Payer FSP is the correct term. If the AccountAddress is opaque (comment above), then it cannot also be used in routing as the information has no meaning in interoperability terms.

> **Regulatory Data Exchange**

Can we avoid that a rouge FSP is not updating other Participants information like changing the fee information for another FSP, by signing data for each Participant? Or should we just trust that all FSPs are behaving correctly, and handle any inconsistencies as they occur?

> **Regulatory Data**
> When preparing a quote, each participant will indicate the data that it requires by putting the names data fields into the `dataRequired` list. It MUST also provide an encryption key that MUST be used by other participants to encrypt any data that they provide.

Is the above concerning the encryption key correct now? The encryption key seems to have been removed from the data model.

> Encryption is done using JSON Web Encryption (RFC 7516).

Same comment as above applies here.

mjbrichards commented:
With apologies for the delay, I'm attaching an example sequence diagram for the discovery phase. I have an example schematic describing the scenario. Will hope to upload quotes and transfers before the end of the week,
![Cross Border Example Schematic 20191123](https://user-images.githubusercontent.com/35496554/71005935-6ce40c00-20dc-11ea-99ee-db8ce77b608e.jpg)


[Cross Implementation Sequence - Discovery.txt](https://github.com/mojaloop/mojaloop-specification/files/3973841/Cross.Implementation.Sequence.-.Discovery.txt)

This is a puml file, by the way...

This makes some controversial assumptions about how things ought to work, but I hope it will provide material for discussion

ehenrka commented:
Thanks @MichaelJBRichards for the detailed sequence diagram! Please find my initial comments below:

- To me it seems much easier to change **PUT /parties/**<_Type_>**/**<_ID_> to return a list of Party elements than having to deal with a header field that says how many callbacks that the client can expect, of which many will be an error callback anyways as the party could not be found by that CNP. The Switch (Scheme A) will then instead be responsible of collecting the results from the different CNPs before sending the callback (which should be similar to how it works today).
- I don't understand why FXPs should be involved in this phase? How do you know based on the **GET /parties/**<_Type_>**/**<_ID_> request that it will involve FX? It might be between two schemes that are using the same currency. It might also be that the FXP is unable to convert from the Payer's currency to the Payee's currency.
- What happens if there are multiple FXPs in the same scheme? Imagine that there would be a FXP F3 in Scheme A, will that mean that the resulting outline route would be for example "FXP F1, FXP F3, CNP 2, FXP F2, BFSP B", or will there be two routes containing the same Party information, "FXP F1, CNP 2, FXP F2, BFSP B" and "FXP F3, CNP 2, FXP F2, BFSP B"?
- Why use FXP F1 as the FSPIOP-Source in the **GET /parties/**<_Type_>**/**<_ID_> as it is the Switch in Scheme A that is requesting the information? (Typo in the note, says FXP X1 instead).
- Would it not be much simpler to just include each FXP that is present in each scheme that is part of the route as additional information? For example the resulting outline route "FXP F1, CNP 2, FXP F2, DFSP B" could be "CNP2 (Scheme A<->Scheme B), DFSP B (Scheme B)", and then additional information regarding Scheme A would include that it has the FXP "FXP F1" (and possibly the "FXP F3" from my comment/question above, and Scheme B has the FXP "FXP F2". The FXPs would then not have to be part of the discovery phase, but they would still be reported back to the client that they can be used during the quote phase as FXPs. The FXPs should not be reuired during the discovery phase.
- Why is a FSPIOP-RequestId necessary? The Switch will send the request to different CNPs, the callback in your flow includes which CNP it originates from in the FSPIOP-Source. The Switch should easily be able to separate the routes using this information without having to add another header field.

Some minor comments that are typos:
- "FSPIP-Destination is CNP 2" FSPIP should be FSPIOP (there are a number of FSPIP, just do a search and replace).
- PUT /parties/ALIAS/B@Somewhere is used in some places where it should be a GET /parties/ALIAS/B@Somewhere
- "FSP X2" sometimes FSP X* is used instead of FSP F*
- "FXP F1, CNP 2, BFSP B", "BFSP B" should be "DFSP B" (there are a number of BFSP instead of DFSP, just do a search and replace)

mjbrichards commented:
Many apologies, Henrik, for having been so long in responding to your (as usual, timely) comments. I've fixed the typos, I think, though I might have added some more. My reason for doing the consolidation was that I was still thinking about the possibility of implementing this in a non-switch implementation; however, I have produced a revised version where the switch does the consolidation. I've also introduced some additional intelligence around the question of whether FXPs add themselves or not; but I concede that they may not fully meet your objections. So here is the revised discovery sequence diagram, together with a quote and a transfer sequence diagram. Let me know what you think



[Cross Implementation Sequence Diagrams.zip](https://github.com/mojaloop/mojaloop-specification/files/4108769/Cross.Implementation.Sequence.Diagrams.zip)

mjbrichards commented:
Curious how jetlag turns one's thoughts to sequence diagrams... This is a further version of the discovery process which (I hope) disambiguates the FX providers and provides each navigable route as a separate entry.

[Cross Implementation Sequence - Discovery V3.zip](https://github.com/mojaloop/mojaloop-specification/files/4111459/Cross.Implementation.Sequence.-.Discovery.V3.zip)

ehenrka commented:
Thanks @mjbrichards for the updated version!

The discovery v3 looks better now in my opinion. Please find my initial comments/questions below. I hope that you can understand where in the sequence diagram a comment originates from, otherwise I will add line numbers from the puml-file.

- `publicKey` is part of the `partyIdInfo` in the **PUT /parties/**<_ID_> callback, I don't think I have seen that one before, could you please point me to some information regarding that?
- There should preferably be a proper error callback instead of a **400 - Bad request** as response if an FXP can't convert to the currency. 
- If there are multiple currencies within one scheme, how will the FXP know that it should convert to XOF based on  **GET /parties/ALIAS/B@Somewhere?currency=ZAR** (please note that there is no body in the **GET** request, so the additional information added by _DFSP B_ will not be part of what is sent to the FXP). It will also not be possible for the FXP add itself to the chain, as there is no body in the **GET /parties** request. I would rather propose another (new) request for communicating with the FXP.
- **GET /parties/ALIAS/B@Somewhere?currency=ZAR** is used as the callback from _FXP F2_, should be **PUT**.
- What does the following mean?: "Am I the last switch? (if so FSPIOP-Destination = party.partyIdInfo\fspId.)"
- "Add the FXP response to the available routes", it was already added by the FXP?
- In the routes in the final **PUT /parties**, the second route does not have ZAR as its transferCurrency to _CNP 2_ (no transferCurrency at all), but the third route does have ZAR as its transferCurrency. Can you please explain why there is a difference (maybe I missed some logic in the middle..). For clarity, I would prefer if there always is a currency.

I will have a look at the quote and transfer sequence diagrams as well, hopefully tomorrow (Tuesday).

mjbrichards commented:
On the publicKey as part of the response: it seemed to me that participants in the chain would not want their information to be visible to each other. I therefore provided a space where the payee DFSP could share a public key which the participants could use to encrypt their contributions so that only the payee DFSP could read them. This would enable the payee DFSP to do its work on preparation of the quote without that information being shared among all parties. I didn't make that thinking very clear in the discovery sequence, but in my defence I think that there is mention of it in the quote sequence diagram. However, we could always decide to omit this functionality if we thought it unncessary.

mjbrichards commented:
On the 400 return, I just wanted a quick way for the FXP to notify the switch that it hadn't added itself to the chain because it couldn't perform the currency conversion required. This isn't an error as such, and of course it would be possible for the switch to infer it from the state of the participant list if it wanted or needed to. So I'm probably happy just for it to be a 200...

Fixed the GET/PUT typo

The FXP is included on the PUT, not the GET. At this stage, the message does have a body, and comparing its **accounts** attribute with the original query string, which contains the currency in which the payer wants to pay, should enable  the switch to determine whether currency conversion is required or not. This is probably a confusion introduced by my GET/PUT typos

The last switch thing is a leftover. I have removed it.

On the FXP response: the FXP adds itself to the single **route** object that was passed to it by the switch (it shouldn't know what other FXP routes there might be...). The switch then adds the **route** object returned by the FXP to its **routes** collection, since there is now a new route which contains the FXP.

On the final point: CNP 3 contains a currency because, in our example, it can only receive in ZAR (and therefore currency conversion cannot be done in scheme A.) CNP 2 does not contain a currency because it can receive either in ZAR or in XOF, so we allow a route where FX happens in scheme A as well as one where FX happens in scheme B. We could ask the switch to set the currencies in all cases, but this could not be done by the CNP itself...

I've made some changes to the sequence diagram to reflect my comments, but will delay uploading a new version until the discussion is more complete

ehenrka commented:
- **Regarding `publicKey`**: From your description this is the Payee FSP's public key, then this should be preferably be placed somewhere else than in the `partyIdInfo` type. This type is used for describing how the party (the Payee in this case) is identified. It is not meant for general FSP information more than the ID of the FSP that the party is located in. My direct assumption from the structure was that this was a public key of the of the party, not the FSP.
- **Regarding the 400 response**: Using a `400` (or other similar error response) would introduce a synchronous behaviour into an otherwise asynchronous API. The FSP would need to check which currencies that it supports conversion between, which would take some more time than just responding with a `202` and then sending a callback. The behaviour needs to follow the rest of the API. This is similar to how for example a Payee FSP can return an error callback with error code `5106` if the Payee does not support the currency.

> The FXP is included on the PUT, not the GET. At this stage, the message does have a body, and comparing its accounts attribute with the original query string, which contains the currency in which the payer wants to pay, should enable the switch to determine whether currency conversion is required or not.
- Sorry, I don't understand the above. The request sent to from _Scheme C_ to _FXP F2_ is **GET /parties/ALIAS/B@Somewhere?currency=ZAR**? On just this **GET** request (which does not have a body), the FXP is meant to be able to know that it should perform a ZAR-XOF conversion. ZAR is included in the URI of the **GET** request, but not XOF. Some of our operators use multiple currencies within the same system, so relying on that XOF is used within that Scheme is not enough.

- **Regarding the FXP response**: Ah, now I understand, that points to what happens in the note that follows directly afterwards. Thanks.
- **Regarding the final point**: I think it would be beneficial for clarity to include a list of all the accepted currencies if there are more than one, instead of leaving it blank.

ehenrka commented:
Regarding the quote in my response which I do not understand, do you mean that it should have been a **PUT** request (which is always callback in the API) sent to the FXP  in the sequence diagram instead of a **GET** request? I'm sorry, but that would go against the core architecture of the API (please see Section 3.2.3 for more information), something I would not allow.

ehenrka commented:
Please find my initial comments regarding the quote flow below. The numbers should reflect the numbering in the sequence diagram.

- **2**: Why is the amount moved to the Payer? I don't like changing existing data models without any apparent reason, since all changes are costly for existing implementers. The following text from the flow don't provide much background. Maybe it is from some other internal discussion:
> I still like the idea of having a send amount and a receive amount
in the request. Whichever amount.amount is blank is the variable
one.If neither is blank, the FXP makes an assumption
- **7**: FXP F1's callback seems to be a **POST /quotes** instead of a **PUT /quotes/**<_ID_>? The quote object is already created at the switch. 
- **18**: In the **PUT /quotes/**<_ID_> from the DFSP B, where is the information regarding the Payee FSP's fee? 800 XOF seems to have just disappeared.
- **25**: Seems more approriate to use a **PATCH /quotes/**<_ID_> (without any callback in nr **28**) here since you are updating an already created quote. The Switch should know that the DFSP A is the final destination without having to ask the FXP. I would like the FXP to be as "dumb" as possible. Leave the routing to the Switch/CNP instead and let the FXP only handle the currency conversion, nothing else.
- CNP 3 is never part of the sequence, but mentioned in the beginning that it will charge 150 ZAR to perform the transfer.

ehenrka commented:
Please find my initial comments regarding the transfer flow below. The numbers should reflect the numbering in the sequence diagram.

- **5**: "Please execute this transfer", what request is sent?
- **Note after 7**: Please avoid using the FXP for routing (see my comment regarding keeping the FXP dumb in the quote phase).
- **8**: Why is the callback a **POST /transfers** instead of a **PUT**?
- **Between 17 and 18**: The Switch in Scheme C should also reserve funds between CNP 2 and DFSP - **35 and 39**: These are both **PUT /transfers**

mjbrichards commented:
On parties, here is a revised version which I think addresses Henrik's architectural concerns. I will respond to Henrik's comments on quotes and transfers shortly.
[Cross Implementation Sequence - Discovery V4.zip](https://github.com/mojaloop/mojaloop-specification/files/4153199/Cross.Implementation.Sequence.-.Discovery.V4.zip)

mjbrichards commented:
On quotes:

In this model, 7 should be a POST because we're still forwarding the quote request through the participants and collecting their information. When it gets to the payee we'll return via a series of PUTs

I'll check through the fees when I have a bit more time.

By 25, we're just passing the completed quote back, not updating it (as I hoped was clearer from the fact that there were no annotations.) There didn't seem to me to be any particular advantage in having the switch make routing decisions, since in this model it doesn't need to be involved at all, except for the standard message forwarding. The FXPs and CNPs will need to have this functionality already, so I don't see any real point in getting the switch to intervene.

I'll take the CNP3 stuff out.

mjbrichards commented:
On transfers:

Note after 7: the transfer needs to go via the FXP because a) the FXP needs to reserve its ZAR funds and b) it may require information from the payer in order to approve the transaction.

After 7: again, it's a POST because we're still outbound. It's a new transfer with a different transfer ID and re-signed by FXP 2. Transaction ID is the same, though.

Will add reservation of funds in Scheme C
Not sure what the point is on 35 and 39: they are PUTs, are you saying they shouldn't be?

mjbrichards commented:
The cross-network presentation Adrian and I gave at the convening is at https://github.com/mojaloop/documentation-artifacts/blob/master/presentations/January%202020%20OSS%20Community%20Session/Cross%20Network%20Design.pdf

ehenrka commented:
> On parties, here is a revised version which I think addresses Henrik's architectural concerns. I will respond to Henrik's comments on quotes and transfers shortly.

Please find my initial comments below regarding this new version.

- What request is the following and what is the body? "Can you do conversion for the following?"
- "Error 3211..", what error callback is sent here?
- The callback to _Scheme A_ from _FXP F1_ should not be a **GET /parties/ALIAS/B@Somewhere?currency=ZAR**.
- Can you please explain the purpose in asking the FXP when you only know about one currency as you don't even know if there is going to be an FX conversion needed? In most cases, there will be one currency within that scheme, and if there is an FXP within that scheme then of course the FXP will support converting that currency. Why not just ask the available FXPs on the way back when you know that there will be an FX conversion needed? The FXPs are being asked multiple times for no apparent reason.
- Related to the above, and related to my earlier comments regarding quotes and transfers about keeping the FXP as dumb as possible, it would be simpler if the message sent to the FXP was a simple **GET /fx/**<_fromCurrency_>**/**<_toCurrency_> or similar, and the FXP returns the callback **PUT /fx/**<_fromCurrency_>**/**<_toCurrency_> . The routing is then handled by the Switch (the Switch would add the FXP to the chain), the implementation in the FXP is easier, the handling is easier to explain in the documentation, the Switch can cache that the FXP can handle this currency conversion, and the messaging would align with the existing architecture. The FXP should not be interested in things like the payer/payee's name at this stage, it will get this information later in the quote if it needs to perform any sanction screening.

I will wait with the updated quote/transfer comments until we have settled something regarding the discovery phase, so that its possible to keep track of this.

mjbrichards commented:
Thanks as usual, Henrik, for your timely responses. I added the actual call to the two instances where the colloquial question was asked: **GET /parties/ALIAS/B@Somewhere?currency=ZAR**

3211 is a new error code that I invented. I didn't want to use 5106, as you suggested, because this isn't a payee error; but I'm not going to go to the stake for it...

In general, I think there is an underlying tension between different fundamental views of how this should be architected, and that's causing a number of more superficial issues. My approach has been to try to minimise the part of the switch in all of this, and to do as much of the work as possible in the FXP, the CNP and the payer. I've also tried, as far as possible, to minimise changes to the API. The alternative view is that we should do as much of the work as possible in the switch. I don't think that I have a very strong attachment to either, so what I'll do is construct an alternative version which I hope will correspond more closely with the switch-centric view of the world. We could then be clearer about the alternative approaches, rather than trying to modify one approach piecemeal.

One thing that I might do, if it seems reasonable to you, is to assume that we will ask FXPs to register with the ALS each currency pair that they support. The switch could then simply obtain from the ALS a list of FXPs which support the required currency pair via a **GET /parties/FXP/ZARXOF**, or similar. Let me know if that seems reasonable to you.

ehenrka commented:
> I added the actual call to the two instances where the colloquial question was asked: GET /parties/ALIAS/B@Somewhere?currency=ZAR

Can you please reply to my question regarding why you need to ask the FXP in this stage as you only know one of the currencies (ZAR in this case)? Please see my previous comment (https://github.com/mojaloop/mojaloop-specification/pull/22#issuecomment-582049720), bullet number 4 for more information.

> 3211 is a new error code that I invented. I didn't want to use 5106, as you suggested, because this isn't a payee error; but I'm not going to go to the stake for it...

Yes, I understood that it was a new error code, as it was a 4-digit code. What I asked about was which error callback (the URI of the HTTP request) that you meant to sent, as there was only an error code. I assume that it is **PUT /parties/ALIAS/B@Somewhere?currency=ZAR/error**? In this case, the Switch will either receive a **PUT** (if it can't handle the currency) or a new **GET** (if it can handle the currency), which seems strange to me and does not follow the current architecture.

> My approach has been to try to minimise the part of the switch in all of this, and to do as much of the work as possible in the FXP, the CNP and the payer.

Can you please explain why your approach is that it should be done in the FXP instead of the Switch? To me, an FXP is just a currency converter. A Switch handles the internal routing within the scheme, and a CNP handles the routing between two schemes.

> One thing that I might do, if it seems reasonable to you, is to assume that we will ask FXPs to register with the ALS each currency pair that they support. The switch could then simply obtain from the ALS a list of FXPs which support the required currency pair via a GET /parties/FXP/ZARXOF, or similar. Let me know if that seems reasonable to you.

Yes, it is very unlikely that the currency pairs that an FXP can handle will change dynamically. Registering them somewhere is perfectly logical and fine by me. My only objection is using **GET /parties/** for this, as you are not requesting information regarding a party. You are asking the FXP if it can perform FX for the provided currency pair, why my suggestion was to use **GET /fx/**<_fromCurrency_>/<_toCurrency_>.

mjbrichards commented:
New version which does the FX conversion routing according (I hope) to Henrik's suggestion...:

[Cross Implementation Sequence - Discovery V5.zip](https://github.com/mojaloop/mojaloop-specification/files/4181380/Cross.Implementation.Sequence.-.Discovery.V5.zip)

ehenrka commented:
Thanks @mjbrichards for the updated version! Please find my comments regarding the updated version below:

- Have you thought anything about receive amount (i.e. the currency is the Payee's currency)? It says "Otherwise, same format as at present." in the note, does that mean it would be just **GET /parties/ALIAS/B@Somewhere**? The other participants would then not know that Ayesha's currency is ZAR, as such it would not be possible to know that currency conversion is necessary or not (if I have not missed something..).
- **GET /parties/CNP** seems a bit odd, as resource **/parties** is for getting information about a Party, not a CNP. I don't see any benefit of having CNP under **/parties** compared to having a separate resource for CNPs. Please compare for example existing separation between **/parties** and **/participants**, I would like us to follow the same pattern here instead of introducing something else. This is also related to comment regarding **GET /parties/FXP/...**.
- There is a note after **GET /parties/CNP** showing a data model, i presume that is a skeleton for the later callback (the **PUT /parties**)?
- There is a question in the flow "Is currency conversion required?" and then "Do nothing". I guess you forgot to add some alternative here?
- "GET /parties/FXP/ZARXOA", please see my last bullet in comment https://github.com/mojaloop/mojaloop-specification/pull/22#issuecomment-582550965. I would like that the request **GET /fx/<fromCurrency>/<toCurrency>** could be sent either to the ALS (which should preferably be renamed as it now knows more than just accounts) or directly to the FXP, similar to **GET /parties** which can be sent to the ALS or directly to the FSP.
- I don't really understand why the callback from the FXP should look like the **PUT /parties** callback. You are replying that the FXP can handle the FX conversion, it has nothing to do with a Party.
- I think it would be beneficial if the participant information for the FXP could include the converted currency as well. Currently the route information looks like:
```
			"route": {
				"participantList": [
					"participant" {
						"fspId": "CNP 2",
						"transferCurrency": "ZAR"
					},
					"participant" {
						"fspId": "FXP 2",
						"transferCurrency": "XOF"
					}
				]
			},
```
Something like the below would give the information that the FSP with fspId _FXP 2_ can perform a currency conversion from ZAR to XOF in the route. 
```
			"route": {
				"participantList": [
					"participant" {
						"fspId": "CNP 2",
						"transferCurrency": "ZAR"
					},
					"participant" {
						"fspId": "FXP 2",
						"transferCurrency": "ZAR",
						"convertedCurrency": "XOF",
					}
				]
			},
```

mjbrichards commented:
We are currently looking at ways of returning the accounts associated with a particular identifier to support account linking for PISPs. It would be helpful if we could agree the extensions to **GET /parties** proposed by Adrian to include account information. Can we raise this as a separate issue?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mojaloop/mojaloop-specification?pullRequest=22) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/mojaloop/mojaloop-specification?pullRequest=22) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: elnyry-sam-k<br/>:x: adrianhopebailie<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/mojaloop/mojaloop-specification?pullRequest=22) it.</sub>

elnyry-sam-k reviewed (approved):
approving PR to add file to the project - doesn't indicate approval of the proposal itself.. will add the file as reference to open issues
